File: Python/_0003_Longest_Substring_Without_Repeating_Characters/Solution.py
# coding=utf-8


# 执行用时：80 ms, 在所有 Python3 提交中击败了56.19%的用户
# 内存消耗：14.8 MB, 在所有 Python3 提交中击败了94.00%的用户
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        f = [0 for _ in range(256)]
        j = res = 0
        for i in range(len(s)):
            f[ord(s[i])] += 1
            while f[ord(s[i])] > 1:
                f[ord(s[j])] -= 1
                j += 1
            res = max(res, i - j + 1)
        return res


if __name__ == "__main__":
    print(Solution().lengthOfLongestSubstring("abcabcbb"))

File: Java/src/_0003_Longest_Substring_Without_Repeating_Characters/Solution.java
package _0003_Longest_Substring_Without_Repeating_Characters;

/**
 * 时间复杂度：O(n)
 * 空间复杂度：O(1)
 */
public class Solution {

    public int lengthOfLongestSubstring(String s) {

        char[] cs = s.toCharArray();

        int[] f = new int[128];  // 表示窗口[j, i]中字符出现的频率
        int res = 0;
        for (int i = 0, j = 0; i < cs.length; i++) {
            f[cs[i]]++;
            while (f[cs[i]] > 1) f[cs[j++]]--;
            res = Math.max(res, i - j + 1);
        }
        return res;
    }

    public static void main(String[] args) {

        System.out.println((new Solution()).lengthOfLongestSubstring("abcabcbb"));
    }
}
